### PR TITLE
[Object][Wasm] Move WasmSymbolInfo directly into WasmSymbol (NFC)

### PR DESCRIPTION
**Pull request by dschuff**
Move the WasmSymbolInfos from their own vector on the WasmLinkingData directly into the WasmSymbol object. Removing the const-ref to an external object allows the vector of WasmSymbols to be safely expanded/reallocated; generating symbol info from the name section will require this, as the numbers of function and data segment names are  stored separately.

This is a step toward generating symbol information from name sections for #76107

### CODE REVIEW

**Comment by dschuff** [comment]
@sbc100 this is a prototype PR to move the `WasmSymbolInfo`s from their own vector on the `WasmLinkingData` directly into the `WasmSymbol` object. Removing the const-ref to an external object allows the vector of `WasmSymbol`s to be safely expanded/reallocated (which we'd need to do if we want to synthesize symbols from export and name sections together). 
It can maybe be further refined, but mostly I'm wondering if you recall why the organization was this way in the first place. The `WasymSymbolInfo`s never seem to be accessed other than via the `WasmSymbol` itself, so this ended up being a very minimal change.

**Comment by dschuff** [comment]
We just discussed live and decided that we don't need to combine symbol info from the export and name sections, that the names from the name section can take precedence. However I realized that we will still need to fix how these are stored because the name section does not declare the total number of names up front (instead it has subsections for each name kind). So assuming we want symbols for both function and data section names, we can't allocate the vectors up front. So we'll either need to change the data representation as this PR does, or do a preparse step to skip ahead in the section and find the total name count before reading the names.

**Comment by dschuff** [comment]
Here's a slightly cleaned-up version; WDYT?

**Review by sbc100 — APPROVED** [review]

File: lld/wasm/InputFiles.cpp [hunk 1]
```diff
@@ -322,20 +322,20 @@ void ObjFile::addLegacyIndirectFunctionTableIfNeeded(
...
-  info->Flags |= WASM_SYMBOL_NO_STRIP;
-  info->ElementIndex = 0;
-  LLVM_DEBUG(dbgs() << "Synthesizing symbol for table import: " << info->Name
+  WasmSymbolInfo info;
+  info.Name = tableImport->Field;
+  info.Kind = WASM_SYMBOL_TYPE_TABLE;
+  info.ImportModule = tableImport->Module;
+  info.ImportName = tableImport->Field;
+  info.Flags = WASM_SYMBOL_UNDEFINED;
+  info.Flags |= WASM_SYMBOL_NO_STRIP;
```
Comment (sbc100) [inline lld/wasm/InputFiles.cpp:331]: Combine the above two lines?

**Pull request merged**